addon/mixins/stateful.js

```diff
@@ -0,0 +1,299 @@
+  }
+  if (!isThenable(ret)) {
+    ret = Ember.RSVP.resolve();
+  }
+  return ret;
+}
+
+function isReservedProperty(name) {
+  return ['_try', '_catch', '_finally', '_default'].includes(name)
+}
+
+function isStateHash(name, obj) {
+  return !isReservedProperty(name) && typeof(obj) !== 'function';
+}
+
+function isAction(name, obj) {
+  return !isReservedProperty(name) && typeof(obj) === 'function';
+}
+
+function getSuperStateName(stateName) {
+  if (stateName === '_root') {
+    return null;
+  }
+  if (!stateName.includes('.')) {
+    return '_root';
+  }
+  return stateName.split('.').slice(0, -1).join('.');
+}
+
+function getStateTaskName(stateName) {
+  if (stateName === '_root') {
+    return '_state_root';
+  }
+  return `_state_${stateName.split('.').join('_')}`;
+}
+
+function getStateTaskGroupPropertyName(stateName) {
+  if (stateName === '_root') {
+    return '_taskgroup_root';
+  }
+  return `_taskgroup_${stateName.split('.').join('_')}`;
+}
+
+function isStateTaskName(name) {
+  return name.startsWith('_state_');
+}
+
+function findSubStates(stateHash) {
+  return Object.entries(stateHash).filter(([name, val]) => isStateHash(name, val));
+}
+
+function findDefaultSubState([stateName, stateHash]) {
+  const subStates = findSubStates(stateHash);
+  if (subStates.length === 0) {
+    return null;
+  }
+  if (subStates.length === 1) {
+    const [subState] = subStates;
+    return subState;
+  }
+  const defaultSubStates = subStates
+    .filter(([, hash]) => hash._default === true)
+  if (defaultSubStates.length === 0) {
+    throw new Error(ERRORS.DEFAULT_SUBSTATE_NOT_DEFINED(stateName))
+  }
+  if (defaultSubStates.length > 1) {
+    throw new Error(ERRORS.MULTIPLE_DEFAULT_SUBSTATES(stateName, defaultSubStates));
+  }
+  const [defaultSubState] = defaultSubStates;
+  return defaultSubState;
+}
+
+function getFullStateName(stateName, fullSuperStateName) {
+  return fullSuperStateName === '_root'
+    ? stateName
+    : [fullSuperStateName, stateName].join('.');
+}
+
+function getStateNameFromStateTaskName(stateTaskName) {
+  if (stateTaskName.endsWith('_root')){
+    return '_root';
+  }
+  const [, state] = stateTaskName.split('_state_');
+  return state.split('_').join('.');
+}
+
+/**
+ * @see https://github.com/emberjs/ember.js/blob/v2.14.1/packages/ember-views/lib/mixins/action_support.js
+ */
+export default Ember.Mixin.create(Ember.Evented, {
+
+  stateTaskNames: computed(function () {
+    return Object.keys(this)
+      .filter(isStateTaskName)
+      .map(key => ({
+        key,
+        val: this.get(key),
+      }))
+      .map(x => x.key)
+      .map(x => (x === '_state_root' ? '_state' : x)) // state_root is actually the topmost state
+      .map(x => x.split('_'))
+      .sort((x, y) => x.length < y.length)
+      .map(x => x.join('_'))
+      .map(x => (x === '_state' ? '_state_root' : x)); // bring back proper name for state_root
+  }),
+
+  stateNames: computed('stateTaskNames', function() {
```

Comment:
is this (public) property on the object absolutely necessary?

Reply:
will hide